In [25]:
import pandas as pd
import numpy as np

Frequency information is stored in the collection of __DatetimeIndex__ object.

In [26]:
dates = pd.date_range(start = '01-01-2017', periods = 500)
dates

DatetimeIndex(['2017-01-01', '2017-01-02', '2017-01-03', '2017-01-04',
               '2017-01-05', '2017-01-06', '2017-01-07', '2017-01-08',
               '2017-01-09', '2017-01-10',
               ...
               '2018-05-06', '2018-05-07', '2018-05-08', '2018-05-09',
               '2018-05-10', '2018-05-11', '2018-05-12', '2018-05-13',
               '2018-05-14', '2018-05-15'],
              dtype='datetime64[ns]', length=500, freq='D')

In [27]:
TT = pd.Series(range(1,501,1), index = dates)
TT.tail()

2018-05-11    496
2018-05-12    497
2018-05-13    498
2018-05-14    499
2018-05-15    500
Freq: D, dtype: int32

# Frequency Indicator

__Freq indicator__ disappeared after modification


In [28]:
TT_temp = TT.copy(deep = True)
TT_temp[pd.to_datetime('05-16-2019')] = 21
TT_temp.tail()

2018-05-12    497
2018-05-13    498
2018-05-14    499
2018-05-15    500
2019-05-16     21
dtype: int64

# Business Calendar

## Grouping Mechanism

_Same as retime/synchronize grouping_ 

In [29]:
TT_Business = TT.resample('BA-DEC').mean()
TT_Business

2017-12-29    182
2018-12-31    432
Freq: BA-DEC, dtype: int32

In [30]:
TT_Business = TT.resample('A-DEC').mean()  # Only weekends are ignored as non-business dates.
TT_Business

2017-12-31    183
2018-12-31    433
Freq: A-DEC, dtype: int32

__Observation__:
- Weekends data are not _ignored/removed_, instead, they are grouped in each bin.
- Non-business dates are grouped into last valid business date.
    - Example: 01-01-2017 is Sunday, so 01-01-2017 and 12-30-2017 are actually grouped into the group 12-29-2017 (Friday)
- 2017-12-30, 2017-12-31 data are actually grouped into 2018 group. 
    - Test: $(364 + 500) / 2= 432$, which is same as the BA frequency resample result.
    - In this case, 364 is the data on 12-30-2017

In [31]:
TT_daily = pd.Series(range(1,4,1), index = pd.date_range(start = '12-30-2016', periods = 3))
TT_daily

2016-12-30    1
2016-12-31    2
2017-01-01    3
Freq: D, dtype: int32

In [32]:
# Grouped into one bin
TT_daily.resample('B').mean()

2016-12-30    2
Freq: B, dtype: int32

# Label Display

In [33]:
TT.head()

2017-01-01    1
2017-01-02    2
2017-01-03    3
2017-01-04    4
2017-01-05    5
Freq: D, dtype: int32

In [34]:
TT.tail()

2018-05-11    496
2018-05-12    497
2018-05-13    498
2018-05-14    499
2018-05-15    500
Freq: D, dtype: int32

In [35]:
TT.resample('M').mean()

2017-01-31     16.0
2017-02-28     45.5
2017-03-31     75.0
2017-04-30    105.5
2017-05-31    136.0
2017-06-30    166.5
2017-07-31    197.0
2017-08-31    228.0
2017-09-30    258.5
2017-10-31    289.0
2017-11-30    319.5
2017-12-31    350.0
2018-01-31    381.0
2018-02-28    410.5
2018-03-31    440.0
2018-04-30    470.5
2018-05-31    493.0
Freq: M, dtype: float64

## Display Left Edge Dates as Labels

In [36]:
TT.resample('M', label = 'left').mean() # Left edge data is not included for calculation by default

2016-12-31     16.0
2017-01-31     45.5
2017-02-28     75.0
2017-03-31    105.5
2017-04-30    136.0
2017-05-31    166.5
2017-06-30    197.0
2017-07-31    228.0
2017-08-31    258.5
2017-09-30    289.0
2017-10-31    319.5
2017-11-30    350.0
2017-12-31    381.0
2018-01-31    410.5
2018-02-28    440.0
2018-03-31    470.5
2018-04-30    493.0
Freq: M, dtype: float64

## Adjust the Resampled Time Labels

In [37]:
TT.resample('M', loffset = '1d').mean()

2017-02-01     16.0
2017-03-01     45.5
2017-04-01     75.0
2017-05-01    105.5
2017-06-01    136.0
2017-07-01    166.5
2017-08-01    197.0
2017-09-01    228.0
2017-10-01    258.5
2017-11-01    289.0
2017-12-01    319.5
2018-01-01    350.0
2018-02-01    381.0
2018-03-01    410.5
2018-04-01    440.0
2018-05-01    470.5
2018-06-01    493.0
Freq: MS, dtype: float64

In [38]:
TT.resample('M', loffset = '1s').mean()

2017-01-31 00:00:01     16.0
2017-02-28 00:00:01     45.5
2017-03-31 00:00:01     75.0
2017-04-30 00:00:01    105.5
2017-05-31 00:00:01    136.0
2017-06-30 00:00:01    166.5
2017-07-31 00:00:01    197.0
2017-08-31 00:00:01    228.0
2017-09-30 00:00:01    258.5
2017-10-31 00:00:01    289.0
2017-11-30 00:00:01    319.5
2017-12-31 00:00:01    350.0
2018-01-31 00:00:01    381.0
2018-02-28 00:00:01    410.5
2018-03-31 00:00:01    440.0
2018-04-30 00:00:01    470.5
2018-05-31 00:00:01    493.0
Freq: M, dtype: float64

# Other Useful Features

## Choose Closed Side of Bin Interval

In [39]:
TT.resample('M', closed = 'left').mean() # [a,b)

2017-01-31     15.5
2017-02-28     44.5
2017-03-31     74.0
2017-04-30    104.5
2017-05-31    135.0
2017-06-30    165.5
2017-07-31    196.0
2017-08-31    227.0
2017-09-30    257.5
2017-10-31    288.0
2017-11-30    318.5
2017-12-31    349.0
2018-01-31    380.0
2018-02-28    409.5
2018-03-31    439.0
2018-04-30    469.5
2018-05-31    492.5
Freq: M, dtype: float64

In [40]:
TT.resample('M').mean() # (a,b]

2017-01-31     16.0
2017-02-28     45.5
2017-03-31     75.0
2017-04-30    105.5
2017-05-31    136.0
2017-06-30    166.5
2017-07-31    197.0
2017-08-31    228.0
2017-09-30    258.5
2017-10-31    289.0
2017-11-30    319.5
2017-12-31    350.0
2018-01-31    381.0
2018-02-28    410.5
2018-03-31    440.0
2018-04-30    470.5
2018-05-31    493.0
Freq: M, dtype: float64

## Set the Origin of the Aggregated Intervals (Intraday)

In [41]:
series = pd.Series(range(10), index=pd.date_range('1/2/2017', periods=10, freq='T'))
series

2017-01-02 00:00:00    0
2017-01-02 00:01:00    1
2017-01-02 00:02:00    2
2017-01-02 00:03:00    3
2017-01-02 00:04:00    4
2017-01-02 00:05:00    5
2017-01-02 00:06:00    6
2017-01-02 00:07:00    7
2017-01-02 00:08:00    8
2017-01-02 00:09:00    9
Freq: T, dtype: int32

In [42]:
series.resample('3T').sum()

2017-01-02 00:00:00     3
2017-01-02 00:03:00    12
2017-01-02 00:06:00    21
2017-01-02 00:09:00     9
Freq: 3T, dtype: int32

In [43]:
# Origin: 2017-01-02 00:00:00 -> 2017-01-02 00:01:00
# 2017-01-02 00:00:00 currently lies in [2017-01-01 23:58:00, 2017-01-01 00:01:00) Interval
series.resample('3T', base = 1).sum()

2017-01-01 23:58:00     0
2017-01-02 00:01:00     6
2017-01-02 00:04:00    15
2017-01-02 00:07:00    24
Freq: 3T, dtype: int32

## Apply Different Aggregation Function on Different Columns

In [44]:
TT = pd.DataFrame(data = np.random.randn(500,2), columns=['Open', 'Volume'],index = dates)
TT.head()

Open    Volume
2017-01-01  0.111683  2.180332
2017-01-02  0.390366 -0.625840
2017-01-03  0.016039  1.319917
2017-01-04 -0.563922  0.037491
2017-01-05 -0.995663  0.925376

In [45]:
TT.resample('M').agg({'Open': np.mean, 'Volume': np.sum})  # No support from TIMETABLE

Open     Volume
2017-01-31 -0.439086   4.246242
2017-02-28 -0.029773   7.056137
2017-03-31  0.018178 -13.912609
2017-04-30 -0.206306  -2.394397
2017-05-31  0.291801   1.130518
2017-06-30  0.208987   1.036354
2017-07-31  0.043218   1.129415
2017-08-31 -0.141247 -11.020090
2017-09-30 -0.038370  -0.334242
2017-10-31 -0.239651  -1.655444
2017-11-30 -0.366803  -2.029371
2017-12-31 -0.082272  -1.310312
2018-01-31  0.240555  -7.560289
2018-02-28  0.128050   3.934745
2018-03-31 -0.024181   6.282551
2018-04-30 -0.160580  -6.709460
2018-05-31  0.195017  -1.308281

In [46]:
TT.resample('M').agg({np.mean, np.std})  # Cartesian Product, Benefit of Multi-Level 

Open              Volume          
                 std      mean       std      mean
2017-01-31  0.950351 -0.439086  0.830079  0.136976
2017-02-28  0.804390 -0.029773  0.870230  0.252005
2017-03-31  1.084659  0.018178  0.933590 -0.448794
2017-04-30  0.982060 -0.206306  0.721963 -0.079813
2017-05-31  1.109038  0.291801  0.938039  0.036468
2017-06-30  1.139125  0.208987  1.269436  0.034545
2017-07-31  0.868497  0.043218  0.983966  0.036433
2017-08-31  0.928313 -0.141247  0.969016 -0.355487
2017-09-30  0.812096 -0.038370  1.065843 -0.011141
2017-10-31  0.798183 -0.239651  0.930682 -0.053401
2017-11-30  0.996823 -0.366803  1.127454 -0.067646
2017-12-31  1.023313 -0.082272  1.184402 -0.042268
2018-01-31  0.793104  0.240555  1.272296 -0.243880
2018-02-28  0.824796  0.128050  1.042549  0.140527
2018-03-31  0.920942 -0.024181  0.696205  0.202663
2018-04-30  0.811331 -0.160580  1.033548 -0.223649
2018-05-31  0.855614  0.195017  1.108684 -0.087219